File: src/test/java/org/snakeyaml/engine/v2/emitter/FlexSimleKeyTest.java
/*
 * Copyright (c) 2018, SnakeYAML
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.snakeyaml.engine.v2.emitter;

import org.junit.jupiter.api.Tag;
import org.junit.jupiter.api.Test;
import org.snakeyaml.engine.v2.api.Dump;
import org.snakeyaml.engine.v2.api.DumpSettings;
import org.snakeyaml.engine.v2.exceptions.YamlEngineException;

import java.util.HashMap;
import java.util.Map;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.fail;

@Tag("fast")
public class FlexSimleKeyTest {

    private final int len = 130;

    @Test
    public void testLongKey() {
        Dump dump = new Dump(createOptions(len));
        Map<String, Object> root = new HashMap();
        Map<String, String> map = new HashMap<>();
        String key = createKey(len);
        map.put(key, "v1");
        root.put("data", map);
        assertEquals("data: {? " + key + "\n  : v1}\n", dump.dumpToString(root));
    }

    @Test
    public void testForceLongKeyToBeImplicit() {
        Dump dump = new Dump(createOptions(len + 10));
        Map<String, Object> root = new HashMap();
        Map<String, String> map = new HashMap<>();
        String key = createKey(len);
        map.put(key, "v1");
        root.put("data", map);
        assertEquals("data: {" + key + ": v1}\n", dump.dumpToString(root));
    }

    @Test
    public void testTooLongKeyLength() {
        try {
            createOptions(1024 + 1);
            fail("Length must be restricted to 1024 chars");
        } catch (YamlEngineException e) {
            assertEquals("The simple key must not span more than 1024 stream characters. See https://yaml.org/spec/1.2/spec.html#id2798057", e.getMessage());
        }
    }

    private DumpSettings createOptions(int len) {
        return DumpSettings.builder().setMaxSimpleKeyLength(len).build();
    }

    private String createKey(int length) {
        StringBuffer outputBuffer = new StringBuffer(length);
        for (int i = 0; i < length; i++) {
            outputBuffer.append("" + (i + 1) % 10);
        }
        String prefix = String.valueOf(length);
        String result = prefix + "_" + outputBuffer.toString().substring(0, length - prefix.length() - 1);
        if (result.length() != length) throw new RuntimeException("It was: " + result.length());
        return result;
    }
}

File: src/commonTest/kotlin/it/krzeminski/snakeyaml/engine/kmp/emitter/FlexSimleKeyTest.kt
package it.krzeminski.snakeyaml.engine.kmp.emitter

import io.kotest.assertions.AssertionErrorBuilder.Companion.fail
import io.kotest.core.spec.style.FunSpec
import io.kotest.matchers.shouldBe
import io.kotest.assertions.throwables.shouldThrowWithMessage
import it.krzeminski.snakeyaml.engine.kmp.api.Dump
import it.krzeminski.snakeyaml.engine.kmp.api.DumpSettings
import it.krzeminski.snakeyaml.engine.kmp.exceptions.YamlEngineException

class FlexSimleKeyTest : FunSpec({
    val len = 130

    test("long key") {
        val dump = Dump(createOptions(len))
        val key = createKey(len)
        val map = mapOf(key to "v1")
        val root = mapOf("data" to map)
        dump.dumpToString(root) shouldBe "data: {? $key\n  : v1}\n"
    }

    test("force long key to be implicit") {
        val dump = Dump(createOptions(len + 10))
        val key = createKey(len)
        val map = mapOf(key to "v1")
        val root = mapOf("data" to map)
        dump.dumpToString(root) shouldBe "data: {$key: v1}\n"
    }

    test("test too long key length") {
        shouldThrowWithMessage<YamlEngineException>(message = "The simple key must not span more than 1024 stream characters. See https://yaml.org/spec/1.2/spec.html#id2798057") {
            createOptions(1024 + 1)
        }
    }
})

private fun createOptions(len: Int): DumpSettings {
    return DumpSettings(maxSimpleKeyLength = len)
}

private fun createKey(length: Int): String {
    val outputBuffer = StringBuilder(length)
    for (i in 0 until length) {
        outputBuffer.append("" + (i + 1) % 10)
    }
    val prefix = length.toString()
    val result = "${prefix}_${outputBuffer.toString().substring(0, length - prefix.length - 1)}"
    if (result.length != length) {
        fail("It was: ${result.length}")
    }
    return result
}
